contracts/Tokens/Prime/Prime.sol

```diff
@@ -0,0 +1,635 @@
+        if (markets[vToken].exists) revert MarketAlreadyExists();
+
+        markets[vToken].rewardIndex = 0;
+        markets[vToken].supplyMultiplier = supplyMultiplier;
+        markets[vToken].borrowMultiplier = borrowMultiplier;
+        markets[vToken].sumOfMembersScore = 0;
+        markets[vToken].exists = true;
+
+        vTokenForAsset[_getUnderlying(vToken)] = vToken;
+
+        allMarkets.push(vToken);
+        _startScoreUpdateRound();
+    }
+
+    /**
+     * @notice Set limits for total tokens that can be mined
+     * @param irrevocableLimit total number of irrevocable tokens that can be minted
+     * @param revocableLimit total number of revocable tokens that can be minted
+     */
+    function setLimit(uint256 irrevocableLimit, uint256 revocableLimit) external {
+        _checkAccessAllowed("setLimit(uint256,uint256)");
+        if (irrevocableLimit < _totalIrrevocable || revocableLimit < _totalRevocable) revert InvalidLimit();
+
+        _revocableLimit = revocableLimit;
+        _irrevocableLimit = irrevocableLimit;
+    }
+
+    /**
+     * @notice Directly issue prime tokens to users
+     * @param isIrrevocable is the tokens being issued is irrevocable
+     * @param users list of address to issue tokens to
+     */
+    function issue(bool isIrrevocable, address[] calldata users) external {
+        _checkAccessAllowed("issue(bool,address[])");
+
+        if (isIrrevocable) {
+            for (uint i = 0; i < users.length; i++) {
+                _mint(true, users[i]);
+                _initializeMarkets(users[i]);
+            }
+        } else {
+            for (uint i = 0; i < users.length; i++) {
+                _mint(false, users[i]);
+                _initializeMarkets(users[i]);
+                delete stakedAt[users[i]];
+            }
+        }
+    }
+
+    /**
+     * @notice Executed by XVSVault whenever user's XVSVault balance changes
+     * @param user the account address whose balance was updated
+     */
+    function xvsUpdated(address user) external {
+        uint256 totalStaked = _xvsBalanceOfUser(user);
+        bool isAccountEligible = isEligible(totalStaked);
+
+        if (tokens[user].exists && !isAccountEligible) {
+            address[] storage _allMarkets = allMarkets;
+            if (tokens[user].isIrrevocable) {
+                for (uint i = 0; i < _allMarkets.length; i++) {
+                    executeBoost(user, _allMarkets[i]);
+                    updateScore(user, _allMarkets[i]);
+                }
+            } else {
+                
+                for (uint i = 0; i < _allMarkets.length; i++) {
+                    executeBoost(user, _allMarkets[i]);
+                    
+                    markets[_allMarkets[i]].sumOfMembersScore = markets[_allMarkets[i]].sumOfMembersScore - interests[_allMarkets[i]][user].score;
+                    interests[_allMarkets[i]][user].score = 0;
+                    interests[_allMarkets[i]][user].rewardIndex = 0;
+                }
```

Comment:
Refactor. I think this for-loop can be moved into the `_burn` function. We invoke `_burn` also in the `burn` function and we are doing the same for-loop (and I think it makes sense to do it every time we burn a token)